katdal/test/test_applycal.py

```diff
@@ -266,6 +267,84 @@ def assert_categorical_data_equal(actual, desired):
         assert_array_equal(a, d)
 
 
+def _non_finite_complex(x):
+    """Identify various kinds of nonfiniteness in complex numbers."""
+    rx = np.real(x)
+    ix = np.imag(x)
+    return (32 * np.isposinf(rx) + 16 * np.isneginf(rx) + 8 * np.isnan(rx)
+            + 4 * np.isposinf(ix) + 2 * np.isneginf(ix) + np.isnan(ix))
+
+
+def assert_array_equal_within_n_ulps(x, y, n=1):
```

Comment:
`assert_array_equal_within_n_ulps` is certainly a complex beast!

It took a while to go through the logic (as I'm sure it must have to write in the first place), but I think I'm happy with it in the end, and appreciate the bit-shifting fun in `_non_finite_complex`. I'm not going to suggest refactoring into smaller subfunctions, since this all reasonably belongs together, simply remark that I hope that whomever inherits the future maintenance of this function is up to the task!

Have you perhaps considered submitting this to `numpy.testing` (i.e. do you see anyone else in need of complex-valued nearness testing?

Reply:
I have. I would build it into `assert_array_max_ulp` though, instead of adding yet another confusing function.

Those functions were originally written by SciPy stalwart David @cournape and hasn't seen much action even though they are recommended.